### ISSUE TITLE
 Cancelling GetHostEntryOrAddressesCoreAsync may cause an unobserved Task Exception event

### ISSUE THREAD

**Issue by baal2000** [dotnet/runtime#61180]
### Description

An application records all `TaskScheduler.UnobservedTaskException` events to make sure that there are no hidden bugs and all asynchronous components' life cycles are controlled.

It are intermittently receives reports of Dns.GetHostEntryOrAddressesCore failing, similar to

```
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Name or service not known)
---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (00000005, 0xFFFDFFFF): Name or service not known
at System.Net.Dns.GetHostEntryOrAddressesCore(String hostName, Boolean justAddresses)
at System.Net.Dns.<>c.<GetHostEntryOrAddressesCoreAsync>b__27_2(Object s)
at System.Threading.Tasks.Task`1.InnerInvoke()
at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of inner exception stack trace ---
```

The failing thread's callstack is rooted in the framework's own code, although it could be difficult to say if it has anything to do with the application or not because the operation is Task-based.
The report source is likely in `GetHostEntryOrAddressesCoreAsync` implementation of a cancellable DNS query when the underlying system does not support cancellation natively. 

[This](https://github.com/dotnet/runtime/blob/7308c4fcc4361c0523c7060d67b5bf4f87ed40e3/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L584) `Task.Run` - like logic:
```
                return RunAsync(static (s, stopwatch) => s switch
                {
                    string h => GetHostAddressesCore(h, AddressFamily.Unspecified, stopwatch),
                    KeyValuePair<string, AddressFamily> t => GetHostAddressesCore(t.Key, t.Value, stopwatch),
                    IPAddress a => GetHostAddressesCore(a, AddressFamily.Unspecified, stopwatch),
                    KeyValuePair<IPAddress, AddressFamily> t => GetHostAddressesCore(t.Key, t.Value, stopwatch),
                    _ => null
                }, asyncState, cancellationToken);
```
may become cancelled  while the synchronous `GetHostAddressesCore` is still waiting for a result. If GetHostAddressesCore fails after `RunAsync` exits due to a signalled `cancellationToken`, an unobserved exception event is going to be triggered.
 

### Reproduction Steps

The test bekow fails with "TestException" message:

```
        [Fact]
        public async Task UnobservedTestAsync()
        {
            Exception? exception = null;
            TaskScheduler.UnobservedTaskException += (_, args) =>
            {
                if (!args.Observed)
                {
                    args.SetObserved();
                }
                exception = args.Exception;
            };

            try
            {
                using var cts = new CancellationTokenSource(100);
                await Task.Run(() =>
                {
                    Thread.Sleep(1000);
                    throw new Exception("TestException");
                }, cts.Token);
            }
            catch (OperationCanceledException)
            {
            }

            await Task.Delay(2000);

            if (exception is not null)
            {
                throw new Exception(exception.Message, exception);
            }
        }
```

### Expected behavior

`GetHostAddressesCore` failure of is ignored if the upstream caller cancels the async wrapper operation. 

### Actual behavior

A `TaskScheduler.UnobservedTaskException` event is logged. This presents a difficult to troubleshoot issue for the Dns API consumers.

### Regression?

I believe that the framework 5.0 and above is affected due to `GetHostAddressesAsync` update to include `CancellationToken` parameter:

from
```
public static Task<IPAddress[]> GetHostAddressesAsync(string hostNameOrAddress)
``` 

to
```
public static Task<IPAddress[]> GetHostAddressesAsync(string hostNameOrAddress, CancellationToken cancellationToken)
```

**Comment by stephentoub** [comment]
> Reproduction Steps

Can you share a repro that includes GetHostXxAsync?  The repro you shared just highlights how UnobservedTaskException is supposed to behave.  If a task generated by the core library is returned to user code and then doesn't have its exception observed, it's correct that UnobservedTaskException would be raised.  What would be unexpected is if the Task returned from GetHostXxAsync properly has any exceptions observed but UnobservedTaskException is still raised.

**Comment by baal2000** [comment]
@stephentoub 
Makes sense, I now see that [RunAsync](https://github.com/dotnet/runtime/blob/7308c4fcc4361c0523c7060d67b5bf4f87ed40e3/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L677) works differently than `Task.Run` although not sure what is expected when cancellationToken becomes cancelled before the wrapped synchronous method finishes with a failure as this behavior is not documented.

Will try to find out.

**Comment by baal2000** [comment]
Another mistake: I have confused 6.0 and 5.0 framework Dns API. There is no cancellation support yet in 5.0 version of [RunAsync](https://github.com/dotnet/runtime/blob/b02e13abab3d6f499751af2ad1ae6333fb4e633d/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L501).

Back to square one. I am going to close the case until have figured out the necessary details.
Thanks.

**Comment by baal2000** [comment]
@stephentoub 
I am still strugging to understand  how this unobserved exception condition om Dns.GetHostEntryOrAddressesCore could happen under the FW 5 as there is no cancellation support that would potentially be able to pre-empt a late failed Task.

Looking at how the synchronous operation is wrapped into a task by RunAsync at https://github.com/dotnet/runtime/blob/b02e13abab3d6f499751af2ad1ae6333fb4e633d/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L480 

I wonder if there is any way (at any point of the application lifecycle) when a Task returned from Task.Factory.StartNew could become completed before `Func<object, TResult> func` completes even with `CancellationToken.None` passed in that would end up in an Unobserved Task event triggered from the Finalizer.

This is FW 5 RunAsync implementation:
```
        private static Task<TResult> RunAsync<TResult>(Func<object, TResult> func, object arg) =>
            Task.Factory.StartNew(func!, arg, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```
Thanks.

**Comment by stephentoub** [comment]
> I wonder if there is any way (at any point of the application lifecycle) when a Task returned from Task.Factory.StartNew could become completed before Func<object, TResult> func completes even with CancellationToken.None passed in that would end up in an Unobserved Task event triggered from the Finalizer.

No.  The CancellationToken passed to StartNew only has any impact before the delegate is invoked.  If cancellation is requested prior to the task actually being run / the delegate invoked, then the task can transition to a Canceled state and the delegate will never be invoked.  If cancellation isn't requested until after the delegate is invoked, cancellation is a nop as far as this Task is concerned.

From what you shared in your initial report, it sounds more like something in the application is calling Dns.GetXxAsync and then isn't paying attention to the returned Task that eventually ends up faulting.  Do you have any "wait until completion or cancellation helpers" in your project?  Consider the Task.WaitAsync method that exists in .NET 6.  If you wrote:
```C#
await Dns.GetXxAsync(..).WaitAsync(cancellationToken);
```
and cancellation was requested prior to the Dns operation completing, the await will wake up and throw the operation canceled exception from the WaitAsync, at which point the application code has abandoned the Task returned from GetXxAsync, and if that Task ends with an exception, the UnobservedTaskException event will be raised.

**Comment by baal2000** [comment]
We do have the await helpers since we moved to async. The tasks they monotor are collected and monitored separately if the helper preempts the upstream await (like WaitAsync in 6.0). When such a Task fails eventually its state is being observed and logged. The failure in question does not look like coming from such a "helped" Task as it remains unobserved. I will continue to investigate and let you know if find anything.

**Comment by baal2000** [comment]
@stephentoub  The "helper class" that causes this could be a part of npgsql library as mentioned in https://github.com/npgsql/npgsql/issues/4149. I could be wrong again, will see.

**Comment by baal2000** [comment]
@stephentoub
Related to unobserved tasks exceptions (not specific to Dns query):

after we upgraded our application to .Net Core 6.0 the test pipeline started failing with
```
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (The operation has timed out.)
---> System.TimeoutException: The operation has timed out.
--- End of inner exception stack trace ---
```

I suspect this is a result of some new framework code using `Task.WaitAsync` and then letting the original Task becomes abandoned.  We will troubleshoot and try create a repo but would appreciate if you could help with ideas. I can create a new issue if it worth doing. Troubleshooting is rather difficult because of the most of context of when the Task in question was created and wrapped into the extension method is lost by the time the error is being reported.

By the way, this issue https://github.com/dotnet/runtime/issues/23878 could be something to consider and prioritize together with WaitAsych implementation if someone was to decide that firing and forgetting is OK on a case by case basis.
